src/xtopcom/xblockmaker/src/xblock_builder.cpp

```diff
@@ -70,22 +70,56 @@ base::xvblock_ptr_t  xunitbuilder_t::create_unit(std::string const& account, uin
     return proposal_block;
 }
 
-base::xvblock_ptr_t  xunitbuilder_t::make_block_v2(const data::xunitstate_ptr_t & unitstate, const data::xblock_consensus_para_t & cs_para){
-    std::string binlog = unitstate->take_binlog();
-    std::string snapshot = unitstate->take_snapshot();
-    if (binlog.empty() || snapshot.empty()) {
-        xerror("xunitbuilder_t::make_block fail-invalid unitstate.");
-        return nullptr;
-    }
+void xunitbuilder_t::make_unitblock_and_unitstate(data::xaccountstate_ptr_t const& accountstate, const data::xblock_consensus_para_t & cs_para, xunit_build_result_t & result) {
+    auto & unitstate = accountstate->get_unitstate();
+    assert(unitstate->is_state_changed());
+    assert(!unitstate->get_bstate()->get_last_block_hash().empty());
+
+    // create unit header firstly for update unitstate information
+    bool is_full_unit = xunitbuilder_t::can_make_full_unit_v2(unitstate->height());
+    std::shared_ptr<data::xunit_build2_t> vblockmaker = std::make_shared<data::xunit_build2_t>(
+        unitstate->get_bstate()->get_account(), unitstate->height(), unitstate->get_bstate()->get_last_block_hash(), is_full_unit, cs_para);    
+
+    // update final unitstate information
+    unitstate->get_bstate()->update_final_block_info(vblockmaker->get_header(), cs_para.get_viewid());
+
+    // make bodypara and create unitblock body
     data::xunit_block_para_t bodypara;
+    std::string binlog = unitstate->take_binlog();
     bodypara.set_binlog(binlog);
+    assert(!binlog.empty());
+  
+    base::enum_xaccountindex_version_t _accountindex_version;
+    std::string _accountindex_state_hash;
+    std::string snapshot;    
```

Comment:
it's easily missunderstood using snapshot for unitstate and unitstate snapshort.